zebra/zebra_nb_state.c

```diff
@@ -147,218 +164,367 @@ lib_interface_zebra_state_mcast_group_get_elem(struct nb_cb_get_elem_args *args)
 	return yang_data_new_ipv4(args->xpath, &vxlan_info->mcast_grp);
 }
 
-const void *lib_vrf_ribs_rib_get_next(struct nb_cb_get_next_args *args)
+const void *lib_vrf_zebra_ribs_rib_get_next(struct nb_cb_get_next_args *args)
 {
-	/* TODO: implement me. */
-	return NULL;
+	struct vrf *vrf = (struct vrf *)args->parent_list_entry;
+	struct zebra_router_table *zrt =
+		(struct zebra_router_table *)args->list_entry;
+
+	struct zebra_vrf *zvrf;
+	afi_t afi;
+	safi_t safi;
+
+	zvrf = zebra_vrf_lookup_by_id(vrf->vrf_id);
+
+	if (args->list_entry == NULL) {
+		afi = AFI_IP;
+		safi = SAFI_UNICAST;
+
+		zrt = zebra_router_find_zrt(zvrf, zvrf->table_id, afi, safi);
+		if (zrt == NULL)
+			return NULL;
+	} else {
+		zrt = RB_NEXT(zebra_router_table_head, zrt);
+		/* vrf_id/ns_id do not match, only walk for the given VRF */
+		while (zrt && zrt->ns_id != zvrf->zns->ns_id)
+			zrt = RB_NEXT(zebra_router_table_head, zrt);
+	}
+
+	return zrt;
 }
 
-int lib_vrf_ribs_rib_get_keys(struct nb_cb_get_keys_args *args)
+int lib_vrf_zebra_ribs_rib_get_keys(struct nb_cb_get_keys_args *args)
 {
-	/* TODO: implement me. */
+	const struct zebra_router_table *zrt = args->list_entry;
+
+	args->keys->num = 2;
+
+	snprintfrr(args->keys->key[0], sizeof(args->keys->key[0]), "%s:%s",
+		   "frr-zebra",
+		   zebra_extract_afi_safi_key(zrt->afi, zrt->safi));
+	snprintfrr(args->keys->key[1], sizeof(args->keys->key[1]), "%" PRIu32,
+		   zrt->tableid);
+
 	return NB_OK;
 }
 
-const void *lib_vrf_ribs_rib_lookup_entry(struct nb_cb_lookup_entry_args *args)
+const void *
+lib_vrf_zebra_ribs_rib_lookup_entry(struct nb_cb_lookup_entry_args *args)
 {
-	/* TODO: implement me. */
-	return NULL;
+	struct vrf *vrf = (struct vrf *)args->parent_list_entry;
+	struct zebra_vrf *zvrf;
+	afi_t afi = AFI_IP;
+	safi_t safi = SAFI_UNICAST;
```

Comment:
`afi` and `safi` must be taken from `args->keys`. Due to this bug I'm seeing some weird multicast routes when entering `show yang operational-data /frr-vrf:lib` :)

Reply:
multicast routes are seen from table afi * safi multicast having connected routes.